npe2/_from_npe1.py

```diff
@@ -403,8 +426,79 @@ def _safe_key(key: str) -> str:
     )
 
 
-def _python_name(object):
-    return f"{object.__module__}:{object.__qualname__}"
+def _python_name(
+    obj: Any, hook: Callable = None, shim_idx: Optional[int] = None
+) -> str:
+    """Get resolvable python name for `obj` returned from an npe1 `hook` implentation.
+
+    Parameters
+    ----------
+    obj : Any
+        a python obj
+    hook : Callable, optional
+        the npe1 hook implementation that returned `obj`, by default None.
+        This is used both to search the module namespace for `obj`, and also
+        in the shim python name if `obj` cannot be found.
+    shim_idx : int, optional
+        If `obj` cannot be found and `shim_idx` is not None, then a shim name.
+        of the form "__npe1shim__.{_python_name(hook)}_{shim_idx}" will be returned.
+        by default None.
+
+    Returns
+    -------
+    str
+       a string that can be imported with npe2.manifest.utils.import_python_name
+
+    Raises
+    ------
+    AttributeError
+        If a resolvable string cannot be found
+    """
+    obj_name: Optional[str] = None
+    mod_name: Optional[str] = None
+    # first, check the global namespace of the module where the hook was declared
+    # if we find `obj` itself, we can just use it.
+    if hasattr(hook, "__module__"):
+        hook_mod = sys.modules.get(hook.__module__)
+        if hook_mod:
+            for local_name, _obj in vars(hook_mod).items():
+                if _obj is obj:
+                    obj_name = local_name
+                    mod_name = hook_mod.__name__
+                    break
+
+    # trick if it's a magic_factory
+    if isinstance(obj, magicgui._magicgui.MagicFactory):
+        f = obj.keywords.get("function")
+        if f:
+            v = getattr(f, "__globals__", {}).get(getattr(f, "__name__", ""))
+            if v is obj:
+                mod_name = f.__module__
+                obj_name = f.__qualname__
+
+    # if that didn't work get the qualname of the object
+    # and, if it's not a locally defined qualname, get the name of the module
+    # in which it is defined
+    if not (mod_name and obj_name):
+        obj_name = getattr(obj, "__qualname__", "")
+        if obj_name and "<locals>" not in obj_name:
+            mod = inspect.getmodule(obj) or inspect.getmodule(hook)
+            if mod:
+                mod_name = mod.__name__
+
+    if not (mod_name and obj_name) and (hook and shim_idx is not None):
+        # we weren't able to resolve an absolute name... if we are shimming, then we
+        # can create a special py_name of the form `__npe1shim__.hookfunction_idx`
+        return f"{SHIM_NAME_PREFIX}{_python_name(hook)}_{shim_idx}"
```

Comment:
maybe this gets hit by later pr's but I'm not seeing test coverage here. This means in turn that `_utils._import_npe1_shim` doesn't get called during tests.  Mostly just noting this here so I'll remember to double check later.